aineko/extras/connectors/rest.py

```diff
@@ -0,0 +1,152 @@
+    url: str
+    headers: Optional[Dict[str, Any]] = None
+    data: Optional[Dict[str, Any]] = None
+    poll_interval: int = 5
+    poll_throttle: float = 0.1
+    max_retries: int = 30
+    metadata: Optional[Dict[str, Any]] = None
+    retry_sleep: float = 5
+
+    @field_validator("url")
+    @classmethod
+    def supported_url(cls, url: str) -> str:  # pylint: disable=no-self-argument
+        """Validates that the url is a valid HTTPS URL."""
+        if not (url.startswith("https://") or url.startswith("http://")):
+            raise ValueError(
+                "Invalid url provided to HTTPS params. "
+                'Expected url to start with "https://". '
+                f"Provided url was: {url}"
+            )
+        return url
+
+
+class REST(AbstractNode):
+    """Connects to an REST endpoint via HTTPS."""
+
+    # Poll settings
+    last_poll_time = time.time()
+    retry_count = 0
+
+    def _pre_loop_hook(self, params: dict | None = None) -> None:
+        """Initializes connection to API."""
+        # Cast params to ParamsREST type
+        try:
+            if params is not None:
+                self.rest_params = ParamsREST(**params)
+            else:
+                raise ValueError(
+                    "No params provided to REST connector. "
+                    "Node requires at least a url param."
+                )
+        except Exception as err:  # pylint: disable=broad-except
+            # Cast pydantic validation error to ValueError
+            # so that it can be properly caught by the node
+            # Note: this is required because pydantic errors
+            # are not pickleable
+            raise ValueError(
+                "Failed to cast params to ParamsREST type. "
+                f"The following error occurred: {err}"
+            ) from err
+        self.rest_params.headers = inject_secrets(self.rest_params.headers)
+        self.rest_params.url = inject_secrets(self.rest_params.url)
+        self.rest_params.data = inject_secrets(self.rest_params.data)
+
+        # Create a session
+        self.log(
+            f"Creating new session to REST endpoint {self.rest_params.url}."
+        )
+        self.session = requests.Session()
+
+    def _execute(self, params: dict | None = None) -> None:
+        """Polls and gets data from the HTTPS endpoint."""
+        # Check if it is time to poll
+        if time.time() - self.last_poll_time >= self.rest_params.poll_interval:
+            # Update the last poll time
+            self.last_poll_time = time.time()
+
+            try:
+                # Poll REST api
+                response = self.session.get(
+                    self.rest_params.url,
+                    timeout=self.rest_params.timeout,
+                    headers=self.rest_params.headers,
+                    data=self.rest_params.data,
+                )
+                # Check if the request was successful
+                if response.status_code != 200:
```

Comment:
Should we broaden the scope to none 2XX codes. I think 2XX codes are all success

Reply:
+1

Reply:
would it be worth adding an optional set of allowable expected return codes, which defaults to 2XX?

Reply:
Great idea. Updated.